Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift

```diff
@@ -78,9 +80,34 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
     }
 
     func close() throws {
+        // Signal long-running operation (e.g., populateTargetTrie) to stop
+        self.isShuttingdown.put(true)
+
         try self.stateLock.withLock {
             if case .connected(let db) = self.state {
-                try db.close()
+                do {
+                    try db.close()
+                } catch {
+                    // This could be because another operation is still running. Wait a bit then try again.
+                    let semaphore = DispatchSemaphore(value: 0)
+                    let callback = { (result: Result<Void, Error>) in
+                        // If second attempt fails as well, semaphore will timeout in which case we will throw error.
+                        if case .success = result {
+                            semaphore.signal()
+                        }
+                    }
+                    self.queue.asyncAfter(deadline: .now() + .milliseconds(200)) {
+                        do {
+                            try db.close()
+                            callback(.success(()))
+                        } catch {
+                            callback(.failure(error))
+                        }
+                    }
+                    guard case .success = semaphore.wait(timeout: .now() + .milliseconds(300)) else {
+                        throw StringError("Failed to close database")
+                    }
+                }
```

Comment:
in general the approach seems fine, but the delay here could be fragile - if the DB operation is longer than 200/300ms it will still fail with "db is locked", we could potentially do a recursive "exponential backoff" style thing here which would be safer. alternatively we could try to eliminate the issue using "sqlite3_interrupt" before trying to close again

Reply:
Exponential backoff: https://github.com/apple/swift-package-manager/pull/3225/commits/0a70a65998622dd8dbe3331b51c936a2b0c62d3d